client/modules/IDE/components/Editor/index.jsx

```diff
@@ -141,546 +228,443 @@ class Editor extends React.Component {
+            };
+          },
+          hintOptions
+        );
+      } else if (cm.current.options.mode === 'css') {
+        // CSS
+        CodeMirror.showHint(cm.current, CodeMirror.hint.css, hintOptions);
+      }
+    };
 
-    this._cm.on('keyup', () => {
-      const temp = this.props.t('Editor.KeyUpLineNumber', {
-        lineNumber: parseInt(this._cm.getCursor().line + 1, 10)
+    cm.current.on('keyup', () => {
+      const temp = t('Editor.KeyUpLineNumber', {
+        lineNumber: parseInt(cm.current.getCursor().line + 1, 10)
       });
       document.getElementById('current-line').innerHTML = temp;
     });
 
-    this._cm.on('keydown', (_cm, e) => {
+    cm.current.on('keydown', (cmm, e) => {
       // Show hint
-      const mode = this._cm.getOption('mode');
+      const mode = cmm.getOption('mode');
       if (/^[a-z]$/i.test(e.key) && (mode === 'css' || mode === 'javascript')) {
-        this.showHint(_cm);
+        showHint(cmm);
       }
     });
-
-    this._cm.getWrapperElement().style[
-      'font-size'
-    ] = `${this.props.fontSize}px`;
-
-    this.props.provideController({
-      tidyCode: this.tidyCode,
-      showFind: this.showFind,
-      showReplace: this.showReplace,
-      getContent: this.getContent
-    });
-  }
-
-  componentWillUpdate(nextProps) {
-    // check if files have changed
-    if (this.props.files[0].id !== nextProps.files[0].id) {
-      // then need to make CodeMirror documents
-      this.initializeDocuments(nextProps.files);
-    }
-    if (this.props.files.length !== nextProps.files.length) {
-      this.initializeDocuments(nextProps.files);
+  }, [mergedState.autocompleteHinter]);
+
+  useEffect(() => {
+    if (file.id) {
+      prevProps.current = {
+        ...preferences,
+        ...ide,
+        ...project,
+        ...editorAccessibility,
+        isExpanded: sidebarIsExpanded,
+        files,
+        fileid: file.id,
+        consoleEvents
+      };
     }
-  }
-
-  componentDidUpdate(prevProps) {
-    if (this.props.file.id !== prevProps.file.id) {
-      const oldDoc = this._cm.swapDoc(this._docs[this.props.file.id]);
-      this._docs[prevProps.file.id] = oldDoc;
-      this._cm.focus();
+  }, []);
+
+  useEffect(() => {
+    const handleChange = () => {
+      funcprop.setUnsavedChanges(true);
+      funcprop.hideRuntimeErrorWarning();
+      funcprop.updateFileContent(file.id, cm.current.getValue());
+      if (mergedState.autorefresh && mergedState.isPlaying) {
+        funcprop.clearConsole();
+        funcprop.startSketch();
+      }
+    };
 
-      if (!prevProps.unsavedChanges) {
-        setTimeout(() => this.props.setUnsavedChanges(false), 400);
+    cm.current.on('change', handleChange);
+    return () => {
+      if (cm.current) {
+        cm.current.off('change', handleChange);
       }
-    }
-    if (this.props.fontSize !== prevProps.fontSize) {
-      this._cm.getWrapperElement().style[
+    };
+  }, [file.id, mergedState.autorefresh, mergedState.isPlaying]);
```

Comment:
This looks good ✔️ .  It seems to have all of the necessary dependencies and I love that you are removing the previous `handleChange` when setting the new one.